src/uu/sort/src/numeric_str_cmp.rs

```diff
@@ -0,0 +1,394 @@
+//!
+//! To be able to short-circuit when comparing, [NumInfo] must be passed along with each number
+//! to [numeric_str_cmp]. [NumInfo] is generally obtained by calling [NumInfo::parse] and should be cached.
+//! It is allowed to arbitrarily modify len afterwards, which is equivalent to shifting the decimal point.
+//!
+//! More specifically, len can be understood as an exponent so that the original number is in (1..10)*10^len.
+//! From that follows the constraints of this algorithm: It is able to compare numbers in ±(1*10^[i64::MIN]..10*10^[i64::MAX]).
+
+use std::{cmp::Ordering, ops::Range};
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum Sign {
+    Negative,
+    Positive,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct NumInfo {
+    len: i64,
+    sign: Sign,
+}
+
+pub struct NumInfoParseSettings {
+    pub accept_si_units: bool,
+    pub thousands_separator: Option<char>,
+    pub decimal_pt: Option<char>,
+}
+
+impl Default for NumInfoParseSettings {
+    fn default() -> Self {
+        Self {
+            accept_si_units: false,
+            thousands_separator: None,
+            decimal_pt: Some('.'),
+        }
+    }
+}
+
+impl NumInfo {
+    /// Parse NumInfo for this number.
+    /// Also returns the range of num that should be passed to numeric_str_cmp later
+    pub fn parse(num: &str, parse_settings: NumInfoParseSettings) -> (Self, Range<usize>) {
+        let chars = num
+            .char_indices()
+            .skip_while(|&(_, c)| c == '0' || c.is_whitespace())
+            .filter(|&(_, c)| parse_settings.thousands_separator != Some(c));
+
+        let mut len = -1;
+        let mut had_decimal_pt = false;
+        let mut had_nonzero_digit = false;
+        let mut start = None;
+        let mut sign = Sign::Positive;
+
+        let mut first_char = true;
+
+        for (idx, char) in chars {
+            if first_char && char == '-' {
+                sign = Sign::Negative;
+                continue;
+            }
+            first_char = false;
+
+            if Self::is_invalid_char(char, &mut had_decimal_pt, &parse_settings) {
+                let si_unit = if parse_settings.accept_si_units {
+                    match char {
+                        'K' | 'k' => 3,
+                        'M' => 6,
+                        'G' => 9,
+                        'T' => 12,
+                        'P' => 15,
+                        'E' => 18,
+                        'Z' => 21,
+                        'Y' => 24,
+                        _ => 0,
+                    }
+                } else {
+                    0
+                };
+                return (
+                    NumInfo {
+                        len: len + si_unit,
+                        sign,
+                    },
+                    start.unwrap_or(0)..idx,
+                );
+            }
+            if Some(char) == parse_settings.decimal_pt {
+                continue;
+            }
+            if !had_decimal_pt {
```

Comment:
You will need additional logic when parsing negative numbers. This test case shows the issue:

```
$ printf -- '-01\n-2' | sort -n
-2
-01

$ printf -- '-01\n-2' | ./target/debug/sort -n
-01
-2
```

I would suggest to handle initial '0's inside the loop for all cases instead of skipping them for positive numbers at https://github.com/uutils/coreutils/blob/43cf667faf5a406b405f3f49f88d46d1ac2fdc66/src/uu/sort/src/numeric_str_cmp.rs#L53

Reply:
Thank you, good catch! I will implement your suggestion. I think generally the bug is that I'm parsing the sign after skipping leading zeroes, but it should be the other way around.